zeebe/dynamic-node-id-provider/src/main/java/io/camunda/zeebe/dynamic/nodeid/RepositoryNodeIdProvider.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.dynamic.nodeid;
+
+import io.camunda.zeebe.dynamic.nodeid.repository.NodeIdRepository;
+import io.camunda.zeebe.dynamic.nodeid.repository.NodeIdRepository.StoredLease;
+import io.camunda.zeebe.dynamic.nodeid.repository.NodeIdRepository.StoredLease.Initialized;
+import io.camunda.zeebe.dynamic.nodeid.repository.NodeIdRepository.StoredLease.Uninitialized;
+import io.camunda.zeebe.util.ExponentialBackoff;
+import java.time.Duration;
+import java.time.InstantSource;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RepositoryNodeIdProvider implements NodeIdProvider, AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RepositoryNodeIdProvider.class);
+  private final NodeIdRepository nodeIdRepository;
+  private final InstantSource clock;
+  private volatile StoredLease.Initialized currentLease;
+  private final Duration leaseDuration;
+  private final String taskId;
+  private final Runnable onLeaseFailure;
+  private final ScheduledExecutorService executor;
+  private final ExponentialBackoff backoff;
+  private long currentDelay;
+  private final Duration renewalDelay;
+  private ScheduledFuture<?> renewalTask;
+
+  public RepositoryNodeIdProvider(
+      final NodeIdRepository nodeIdRepository,
+      final InstantSource clock,
+      final Duration expiryDuration,
+      final String taskId,
+      final Runnable onLeaseFailure) {
+    this.nodeIdRepository =
+        Objects.requireNonNull(nodeIdRepository, "nodeIdRepository cannot be null");
+    this.clock = Objects.requireNonNull(clock, "clock cannot be null");
+    leaseDuration = Objects.requireNonNull(expiryDuration, "expiryDuration cannot be null");
+    this.taskId = Objects.requireNonNull(taskId, "taskId cannot be null");
+    this.onLeaseFailure = Objects.requireNonNull(onLeaseFailure, "onLeaseFailure cannot be null");
+    backoff = new ExponentialBackoff(Duration.ofSeconds(1), leaseDuration.dividedBy(2));
+    renewalDelay = leaseDuration.dividedBy(3);
+    currentDelay = 0L;
+    executor = Executors.newSingleThreadScheduledExecutor(r -> new Thread(r, "NodeIdProvider"));
+  }
+
+  @Override
+  public void initialize(final int clusterSize) {
+    nodeIdRepository.initialize(clusterSize);
+    CompletableFuture.runAsync(() -> acquireInitialLease(clusterSize), executor)
```

Comment:
:question: This is not waiting until the lease is acquired. How is it ensured that, the broker do not start before the nodeId is assigned?